### ISSUE TITLE
EventBridge Retry Policy Implementation

### ISSUE THREAD

**Issue by utajum** [localstack/localstack#11630]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Feature description

LocalStack should implement the EventBridge retry policy feature as described in the AWS documentation [1].

Currently, in LocalStack version 3.7.2, failed events are never retried, which differs from the behaviour of the actual AWS EventBridge service.

According to the AWS documentation, EventBridge has a retry policy for events that aren't successfully delivered to their targets. The default behaviour is to retry sending the event for 24 hours and up to 185 times with exponential backoff and jitter. 

Implementing this feature in LocalStack would provide a more accurate simulation of the AWS EventBridge service.

My setup
docker-compose:

```
  localstack:
    extra_hosts:
      - "host.docker.internal:host-gateway"
    image: localstack/localstack:3.7.2
    environment:
      DEBUG: ${DEBUG-}
      DOCKER_HOST: unix:///var/run/docker.sock
      SERVICES: s3,secretsmanager,events
      GATEWAY_LISTEN: "0.0.0.0:4567"
      AWS_ENDPOINT_URL: "http://localstack:4567"
      # PROVIDER_OVERRIDE_EVENTS: v2
```

If I use `PROVIDER_OVERRIDE_EVENTS: v2` EventBridge does not work

Setting `RetryPolicy` on `PutTargetsCommand` does nothing.

I have tested the same code with AWS EventBridge and failed attempts are queued and retried.

[1] https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-rule-retry-policy.html

### Anything else?

https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-rule-retry-policy.html

**Comment by zaingz** [comment]
@utajum, thanks for raising the issue! I’ve been looking into the PROVIDER_OVERRIDE_EVENTS specifically but haven’t been able to reproduce the problem. Could you provide some additional context to help me investigate further?
	•	Does the issue still persist?
	•	Is it specific to version 3.7.2? It’s possible it’s resolved in the latest version.
	•	When you say, “If I use PROVIDER_OVERRIDE_EVENTS: v2 EventBridge does not work,” do you mean the flag isn’t being applied, or that retries aren’t functioning as expected?

**Comment by utajum** [comment]
I'm not sure if this issue still persists, but I experienced the same problem when testing across different versions. Specifically, when setting the PROVIDER_OVERRIDE_EVENTS flag to v2, EventBridge stopped working entirely. Nothing related to EventBridge in LocalStack functioned when this flag was enabled and yet it is stated in the documentation

**Comment by zaingz** [comment]
Thank you for the feedback! To help investigate this further, could you provide some additional details? This would help us understand where exactly EventBridge v2 is failing.

A few suggestions that might help:

1. Could you enable debug logging by setting:
```bash
DEBUG=1
```
This will provide detailed logs about what's happening when EventBridge operations are attempted.

2. I notice you're using the `SERVICES` environment variable. We generally recommend not restricting services this way as it can sometimes cause unexpected behavior with dependent services. Could you try without this setting?

3. Could you also share:
- The exact LocalStack version you're using (`localstack --version`)
- A minimal example of an EventBridge operation that fails
- The resulting debug logs when the operation fails

This information would help us understand whether this is a configuration issue or a deeper problem with the EventBridge v2 implementation.

**Comment by utajum** [comment]
The topic of this issue is not me debugging the PROVIDER_OVERRIDE_EVENTS variable for you. 

At the time of opening this issue, the latest localstack version was 3.7.2,  and in the localstack documentation it was recommended to set `PROVIDER_OVERRIDE_EVENTS: v2`, this is now changed to `v1`


I did some tests now with version 4.0.2

## This config works, but still no Retry Policy implementation (the topic of this issue)
```yaml
    image: localstack/localstack:4.0.2
    environment:
      DEBUG: 1
      DOCKER_HOST: unix:///var/run/docker.sock
      SERVICES: s3,secretsmanager,events
      GATEWAY_LISTEN: "0.0.0.0:4567"
      AWS_ENDPOINT_URL: "http://localstack:4567"
      PROVIDER_OVERRIDE_EVENTS: v1
```


## This config does not work
```yaml
    image: localstack/localstack:4.0.2
    environment:
      DEBUG: 1
      DOCKER_HOST: unix:///var/run/docker.sock
      SERVICES: s3,secretsmanager,events
      GATEWAY_LISTEN: "0.0.0.0:4567"
      AWS_ENDPOINT_URL: "http://localstack:4567"
      #PROVIDER_OVERRIDE_EVENTS: v1 # nor v2
```

With version 3.0.7 there were errors when creating resources, with 4.0.2 payload does not get sent to the target destination connection

#### 4.0.2
```
2024-11-28T13:44:36.436 DEBUG --- [et.reactor-1] l.s.events.target_helper   : Calling EventBridge API destination (state "ACTIVE"): POST https://e2e1-26-217-81-329.ngrok-free.app/api/outgoing-workflows
2024-11-28T13:44:36.440  INFO --- [et.reactor-1] l.services.events.provider : {"ErrorCode": "TargetDeliveryFailure", "ErrorMessage": "Failed to deliver to target staging-test-rule-target: Object of type datetime is not JSON serializable"}
2024-11-28T13:44:36.440  INFO --- [et.reactor-1] l.services.events.provider : {"InfoCode": "InternalInfoEvents at matches_rule", "InfoMessage": "No rules matched for formatted event: {'version': '0', 'id': '45b82ddf-6540-4abb-9059-df752c121ad6', 'detail-type': 'testdetail', 'source': 'test-query', 'account': '000000000000', 'time': datetime.datetime(2024, 11, 28, 13, 44, 36, 427076, tzinfo=datetime.timezone.utc), 'region': 'eu-north-1', 'resources': [], 'detail': {'key1': 'value1', 'key2': 'value2', 'source': 'test-query', 'tenantId': '123456'}, 'event-bus-name': 'staging'}"}
2024-11-28T13:44:36.440  INFO --- [et.reactor-1] localstack.request.aws     : AWS events.PutEvents => 200
```

#### The same payload was used for the tests, AWS works fine

**Comment by zaingz** [comment]
Hi,
I apologize for the confusion in my previous response. You’re right - this issue is about the EventBridge retry policy implementation, not about the provider override settings.
To clarify the current status:
1. EventBridge retry policy feature is indeed not implemented yet - this is on our roadmap
2. From the logs you shared, there seems to be an additional issue with datetime serialization in v2 provider, which is preventing events from reaching targets
For the retry policy implementation:
- The current behavior doesn’t match AWS’s default retry behavior ([24 hours with up to 185 retries with exponential backoff](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-rule-retry-policy.html))
- For the datetime serialization error you’re encountering in 4.0.2:
```json
“ErrorMessage”: “Failed to deliver to target staging-test-rule-target: Object of type datetime is not JSON serializable”
```
This is a separate issue that should be fixed. I have create a fix for this: https://github.com/localstack/localstack/pull/11959

**Comment by joe4dev** [comment]
Hi @utajum, let's re-open this issue for transparency because the retry policy is still a valid feature request.

The "datetime serialization" is fixed and should unblock your use case.

**Comment by joemcbride** [comment]
+1 for adding support for the EventBridge retry policy feature

**Comment by HarshCasper** [comment]
Due to the forthcoming changes in our distribution and development model ([read more](https://blog.localstack.cloud/the-road-ahead-for-localstack/)), this issue tracker is no longer being maintained. We've migrated this issue to our internal tracker and will continue to work on it there. If you'd like to follow up on the status or expedite this, feel free to reach out through one of our support channels:

-   💬 [GitHub Discussions](https://github.com/orgs/localstack/discussions)
-   📖 [Help & Support](https://docs.localstack.cloud/aws/getting-started/help-support/)
